diff --git a/Makefile b/Makefile
index 09d790c..988eedb 100644
--- a/Makefile
+++ b/Makefile
@@ -181,6 +181,7 @@ UPROGS=\
 	_usertests\
 	_wc\
 	_zombie\
+	_trylseek\
 
 fs.img: mkfs README $(UPROGS)
 	./mkfs fs.img README $(UPROGS)
diff --git a/fcntl.h b/fcntl.h
index d565483..f30ec4a 100644
--- a/fcntl.h
+++ b/fcntl.h
@@ -2,3 +2,7 @@
 #define O_WRONLY  0x001
 #define O_RDWR    0x002
 #define O_CREATE  0x200
+
+#define SEEK_SET  0
+#define SEEK_CUR  1
+#define SEEK_END  2
diff --git a/syscall.c b/syscall.c
index ee85261..e72a40f 100644
--- a/syscall.c
+++ b/syscall.c
@@ -92,6 +92,7 @@ extern int sys_fstat(void);
 extern int sys_getpid(void);
 extern int sys_kill(void);
 extern int sys_link(void);
+extern int sys_lseek(void);
 extern int sys_mkdir(void);
 extern int sys_mknod(void);
 extern int sys_open(void);
@@ -126,6 +127,7 @@ static int (*syscalls[])(void) = {
 [SYS_link]    sys_link,
 [SYS_mkdir]   sys_mkdir,
 [SYS_close]   sys_close,
+[SYS_lseek]   sys_lseek,
 };
 
 void
diff --git a/syscall.h b/syscall.h
index bc5f356..93fbf00 100644
--- a/syscall.h
+++ b/syscall.h
@@ -20,3 +20,4 @@
 #define SYS_link   19
 #define SYS_mkdir  20
 #define SYS_close  21
+#define SYS_lseek  22
diff --git a/sysfile.c b/sysfile.c
index bfe61b7..96a2494 100644
--- a/sysfile.c
+++ b/sysfile.c
@@ -90,6 +90,55 @@ sys_write(void)
   return filewrite(f, p, n);
 }
 
+#define MAXOFFSET 0x7fffffff
+
+int
+sys_lseek(void)
+{
+  struct file *f;
+  int offset;
+  int whence;
+  int newoff;
+
+  if(argfd(0, 0, &f) < 0)
+    return -1;
+
+  if(argint(1, &offset) < 0)
+    return -1;
+
+  if(argint(2, &whence) < 0)
+    return -1;
+
+  if(f->type != FD_INODE || f->ip->type != T_FILE)
+    return -1;
+
+  if(whence == SEEK_SET){
+    if(offset < 0)
+      return -1;
+    newoff = offset;
+  } else if(whence == SEEK_CUR){
+    if(offset > 0 && f->off > MAXOFFSET - offset)
+      return -1;
+    if(offset < 0 && offset < -MAXOFFSET)
+      return -1;
+    newoff = f->off + offset;
+    if(newoff < 0)
+      return -1;
+  } else if(whence == SEEK_END){
+    if(offset < 0 && offset < -(int)f->ip->size)
+      return -1;
+    if(offset > 0 && f->ip->size > MAXOFFSET - offset)
+      return -1;
+    newoff = f->ip->size + offset;
+    if(newoff < 0)
+      return -1;
+  } else
+    return -1;
+
+  f->off = newoff;
+  return f->off;
+}
+
 int
 sys_close(void)
 {
diff --git a/trylseek.c b/trylseek.c
new file mode 100644
index 0000000..e5cb730
--- /dev/null
+++ b/trylseek.c
@@ -0,0 +1,40 @@
+#include "types.h"
+#include "fcntl.h"
+#include "stat.h"
+#include "user.h"
+#include "fs.h"
+
+int
+main(int argc, char *argv[])
+{
+  int fd, i, n;
+  char buf[128];
+
+  fd = open(argv[1], O_RDONLY);
+
+  lseek(fd, 10, SEEK_SET);
+  n = read(fd, buf, 5);
+  for(i = 0; i < n; i++)
+    printf(1, "%c", buf[i]);
+  printf(1, "\n");
+
+  lseek(fd, 10, SEEK_CUR);
+  n = read(fd, buf, 5);
+  for(i = 0; i < n; i++)
+    printf(1, "%c", buf[i]);
+  printf(1, "\n");
+
+  lseek(fd, -5, SEEK_END);
+  n = read(fd, buf, 5);
+  for(i = 0; i < n; i++)
+    printf(1, "%c", buf[i]);
+  printf(1, "\n");
+
+  lseek(fd, 5, SEEK_END);
+  n = read(fd, buf, 5);
+  for(i = 0; i < n; i++)
+    printf(1, "%c", buf[i]);
+  printf(1, "\n");
+
+  exit();
+}
diff --git a/user.h b/user.h
index 4f99c52..009f50d 100644
--- a/user.h
+++ b/user.h
@@ -8,6 +8,7 @@ int wait(void);
 int pipe(int*);
 int write(int, const void*, int);
 int read(int, void*, int);
+int lseek(int, int, int);
 int close(int);
 int kill(int);
 int exec(char*, char**);
diff --git a/usys.S b/usys.S
index 8bfd8a1..6244300 100644
--- a/usys.S
+++ b/usys.S
@@ -29,3 +29,4 @@ SYSCALL(getpid)
 SYSCALL(sbrk)
 SYSCALL(sleep)
 SYSCALL(uptime)
+SYSCALL(lseek)
